### ISSUE TITLE
Import h5py fails in serial code after being built with parallel support.

### ISSUE THREAD

**Issue by leonfoks** [h5py/h5py#917]
I have h5py built with parallel support and it is working fine with no issues in parallel.  However, if I try to "import h5py" in a serial run i.e.

```
python -c 'import h5py'
```
instead of
```
srun (or mpirun) --mpi=pmi2 python -c 'import h5py'
```
it fails with the message attached to the bottom of this issue.  The message is related to the mpi library not being able to initialize because it was not run using srun or mpirun.

I believe that the user should still be able to import and run h5py serially despite a parallel install.
Perhaps mpi4py should be initialized when the user opens a file with the mpio driver, and not at import?

OS: Scientific linux
Python: 3.6.0 :: Anaconda 4.3.1
Openmpi: 1.10.2
HDF5: 1.8.18
mpi4py: 2.0.0
h5py: 2.8.0.post0

Thanks!

```
--------------------------------------------------------------------------
It looks like orte_init failed for some reason; your parallel process is
likely to abort.  There are many reasons that a parallel process can
fail during orte_init; some of which are due to configuration or
environment problems.  This failure appears to be an internal failure;
here's some additional information (which may only be relevant to an
Open MPI developer):

  PMI2_Job_GetId failed failed
  --> Returned value (null) (14) instead of ORTE_SUCCESS
--------------------------------------------------------------------------
--------------------------------------------------------------------------
It looks like orte_init failed for some reason; your parallel process is
likely to abort.  There are many reasons that a parallel process can
fail during orte_init; some of which are due to configuration or
environment problems.  This failure appears to be an internal failure;
here's some additional information (which may only be relevant to an
Open MPI developer):

  orte_ess_init failed
  --> Returned value (null) (14) instead of ORTE_SUCCESS
--------------------------------------------------------------------------
--------------------------------------------------------------------------
It looks like MPI_INIT failed for some reason; your parallel process is
likely to abort.  There are many reasons that a parallel process can
fail during MPI_INIT; some of which are due to configuration or environment
problems.  This failure appears to be an internal failure; here's some
additional information (which may only be relevant to an Open MPI
developer):

  ompi_mpi_init: ompi_rte_init failed
  --> Returned "(null)" (14) instead of "Success" (0)
--------------------------------------------------------------------------
*** An error occurred in MPI_Init_thread
*** on a NULL communicator
*** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
***    and potentially your MPI job)
```

**Comment by aragilar** [comment]
@leonfoks Would you be able to get a backtrace?

**Comment by leonfoks** [comment]
I just tried a backtrace with both a try except and the traceback package, and pdb using pdb.set_trace(), but this error causes python to exit before the trace can spit anything out.  Any suggestions to get a backtrace in this case would be great!

Note:
I get the same error in serial if I "from mpi4py import MPI" because it tries to initialize the mpi communicator but cannot.  I can, however, successfully "import mpi4py" and access mpi4py.get_config() etc. in serial.

Sorry if this is not that helpful, I will try any suggestions!

**Comment by aragilar** [comment]
Sorry, I meant a backtrace using a debugger like gdb, see https://wiki.debian.org/HowToGetABacktrace. Even if you don't have the debugging symbols available, there should be enough information in the binaries for gdb to identify in which library the segfault is occurring.

**Comment by boegel** [comment]
I'm seeing something similar, it seems, or at least it looks related.
In my case, `python -c "import h5py"` doesn't fail, but it hangs, and when checking with GDB it seems to be stuck in MPI init, see below.

This is also on a workernode in a SLURM cluster, and I'm only seeing it when `h5py` is being built on top of Intel MPI; it doesn't occur when using Open MPI.

```
(gdb) bt
#0  0x00002b2f87ed9410 in MPID_nem_barrier () at ../../src/mpid/ch3/channels/nemesis/src/mpid_nem_barrier.c:74
#1  0x00002b2f87ed840f in MPIDI_CH3I_Seg_commit (memory=0x2b2f88809818 <PVAR_TIMER_idle+16>, num_local=-1996005376, local_rank=-2004838392) at ../../src/mpid/ch3/channels/nemesis/src/mpid_nem_alloc.c:449
#2  0x00002b2f87ee9431 in MPID_nem_init_ckpt (pg_rank=-2004838376, pg_p=0x2b2f89076000, ckpt_restart=-2004838392, has_parent=-2005061824) at ../../src/mpid/ch3/channels/nemesis/src/mpid_nem_init.c:848
#3  0x00002b2f87cb60d1 in MPIDI_CH3_Init (has_parent=-2004838376, pg_p=0x2b2f89076000, pg_rank=-2004838392) at ../../src/mpid/ch3/channels/nemesis/src/ch3_init.c:125
#4  0x00002b2f87ed48ee in MPID_Init (argc=0x2b2f88809818 <PVAR_TIMER_idle+16>, argv=0x2b2f89076000, requested=-2004838392, provided=0x2b2f887d2f40 <MPID_nem_mem_region>, has_args=0x0, has_env=0x4f)
    at ../../src/mpid/ch3/src/mpid_init.c:1857
#5  0x00002b2f87e7542b in MPIR_Init_thread (argc=0x2b2f88809818 <PVAR_TIMER_idle+16>, argv=0x2b2f89076000, required=-2004838392, provided=0x2b2f887d2f40 <MPID_nem_mem_region>)
    at ../../src/mpi/init/initthread.c:717
#6  0x00002b2f87e730d6 in PMPI_Init_thread (argc=0x2b2f88809818 <PVAR_TIMER_idle+16>, argv=0x2b2f89076000, required=-2004838392, provided=0x2b2f887d2f40 <MPID_nem_mem_region>)
    at ../../src/mpi/init/initthread.c:1061
#7  0x00002b2f88bb8116 in __pyx_f_6mpi4py_3MPI_bootstrap () at src/mpi4py.MPI.c:7034
#8  0x00002b2f88b80eef in initMPI () at src/mpi4py.MPI.c:163327
#9  0x00002b2f7cf06646 in _PyImport_LoadDynamicModule (name=0x2b2f88809818 <PVAR_TIMER_idle+16> "", pathname=0x2b2f89076000 "\001", fp=0x2b2f88809808 <PVAR_TIMER_idle>) at ./Python/importdl.c:53
#10 0x00002b2f7cf03cde in load_module (name=0x2b2f88809818 <PVAR_TIMER_idle+16> "", fp=0x2b2f89076000, pathname=0x2b2f88809808 <PVAR_TIMER_idle> "", type=-2005061824, loader=0x0) at Python/import.c:1937
#11 0x00002b2f7cf030d5 in import_submodule (mod=<optimized out>, subname=<optimized out>, fullname=<optimized out>) at Python/import.c:2725
#12 load_next (mod=<optimized out>, altmod=<optimized out>, p_name=<optimized out>, buf=<optimized out>, p_buflen=<optimized out>) at Python/import.c:2539
#13 import_module_level (name=0x2b2f88809818 <PVAR_TIMER_idle+16> "", globals=0x2b2f89076000, locals=0x2b2f88809808 <PVAR_TIMER_idle>, fromlist=0x2b2f887d2f40 <MPID_nem_mem_region>, level=0)
    at Python/import.c:2256
#14 0x00002b2f7cf057b0 in PyImport_ImportModuleLevel (name=0x2b2f88809818 <PVAR_TIMER_idle+16> "", globals=0x2b2f89076000, locals=0x2b2f88809808 <PVAR_TIMER_idle>,
    fromlist=0x2b2f887d2f40 <MPID_nem_mem_region>, level=0) at Python/import.c:2312
#15 0x00002b2f7cecae7e in builtin___import__ (self=0x2b2f88809818 <PVAR_TIMER_idle+16>, args=0x2b2f89076000, kwds=0x2b2f88809808 <PVAR_TIMER_idle>) at Python/bltinmodule.c:49
#16 0x00002b2f7cdee38a in PyObject_Call (func=0x2b2f88809818 <PVAR_TIMER_idle+16>, arg=0x2b2f89076000, kw=0x2b2f88809808 <PVAR_TIMER_idle>) at Objects/abstract.c:2547
#17 0x00002b2f7cdf45eb in call_function_tail (callable=<optimized out>, args=0x2b2f8079cad0) at Objects/abstract.c:2579
#18 PyObject_CallFunction (callable=0x2b2f88809818 <PVAR_TIMER_idle+16>, format=0x2b2f89076000 "\001") at Objects/abstract.c:2603
#19 0x00002b2f7cf058bf in PyImport_Import (module_name=0x2b2f88809818 <PVAR_TIMER_idle+16>) at Python/import.c:2911
#20 0x00002b2f86cf65bf in __Pyx_ImportModule (name=<optimized out>) at /tmp/vsc40003/easybuild/h5py/2.7.1/intel-2018a-Python-2.7.14/h5py-2.7.1/h5py/_errors.c:4798
#21 __Pyx_ImportType..0 (module_name=0x2b2f88809818 <PVAR_TIMER_idle+16> "", class_name=0x2b2f89076000 "\001", size=47483153586184, strict=-2005061824)
    at /tmp/vsc40003/easybuild/h5py/2.7.1/intel-2018a-Python-2.7.14/h5py-2.7.1/h5py/_errors.c:4821
#22 0x00002b2f86cf8001 in init_errors () at /tmp/vsc40003/easybuild/h5py/2.7.1/intel-2018a-Python-2.7.14/h5py-2.7.1/h5py/_errors.c:2692
#23 0x00002b2f7cf06646 in _PyImport_LoadDynamicModule (name=0x2b2f88809818 <PVAR_TIMER_idle+16> "", pathname=0x2b2f89076000 "\001", fp=0x2b2f88809808 <PVAR_TIMER_idle>) at ./Python/importdl.c:53
#24 0x00002b2f7cf03cde in load_module (name=0x2b2f88809818 <PVAR_TIMER_idle+16> "", fp=0x2b2f89076000, pathname=0x2b2f88809808 <PVAR_TIMER_idle> "", type=-2005061824, loader=0x0) at Python/import.c:1937
#25 0x00002b2f7cf05556 in import_submodule (mod=<optimized out>, subname=<optimized out>, fullname=<optimized out>) at Python/import.c:2725
#26 ensure_fromlist (mod=0x2b2f88809818 <PVAR_TIMER_idle+16>, fromlist=0x2b2f89076000, buf=0x2b2f88809808 <PVAR_TIMER_idle> "", buflen=47483153362752, recursive=0) at Python/import.c:2631
#27 0x00002b2f7cf03298 in import_module_level (name=0x2b2f88809818 <PVAR_TIMER_idle+16> "", globals=0x2b2f89076000, locals=0x2b2f88809808 <PVAR_TIMER_idle>,
    fromlist=0x2b2f887d2f40 <MPID_nem_mem_region>, level=0) at Python/import.c:2293
#28 0x00002b2f7cf057b0 in PyImport_ImportModuleLevel (name=0x2b2f88809818 <PVAR_TIMER_idle+16> "", globals=0x2b2f89076000, locals=0x2b2f88809808 <PVAR_TIMER_idle>,
    fromlist=0x2b2f887d2f40 <MPID_nem_mem_region>, level=0) at Python/import.c:2312
#29 0x00002b2f7cecae7e in builtin___import__ (self=0x2b2f88809818 <PVAR_TIMER_idle+16>, args=0x2b2f89076000, kwds=0x2b2f88809808 <PVAR_TIMER_idle>) at Python/bltinmodule.c:49
#30 0x00002b2f7cdee38a in PyObject_Call (func=0x2b2f88809818 <PVAR_TIMER_idle+16>, arg=0x2b2f89076000, kw=0x2b2f88809808 <PVAR_TIMER_idle>) at Objects/abstract.c:2547
#31 0x00002b2f7ced8d3f in PyEval_CallObjectWithKeywords (func=0x2b2f88809818 <PVAR_TIMER_idle+16>, arg=0x2b2f89076000, kw=0x2b2f88809808 <PVAR_TIMER_idle>) at Python/ceval.c:4226
#32 0x00002b2f7ced4a5d in PyEval_EvalFrameEx (f=0x2b2f88809818 <PVAR_TIMER_idle+16>, throwflag=-1996005376) at Python/ceval.c:2628
#33 0x00002b2f7ced97ed in PyEval_EvalCodeEx (co=0x2b2f88809818 <PVAR_TIMER_idle+16>, globals=0x2b2f89076000, locals=0x2b2f88809808 <PVAR_TIMER_idle>, args=0x2b2f887d2f40 <MPID_nem_mem_region>,
    argcount=0, kws=0x4f, kwcount=0, defs=0x0, defcount=0, closure=0x0) at Python/ceval.c:3589
#34 0x00002b2f7ced92b4 in PyEval_EvalCode (co=0x2b2f88809818 <PVAR_TIMER_idle+16>, globals=0x2b2f89076000, locals=0x2b2f88809808 <PVAR_TIMER_idle>) at Python/ceval.c:669
#35 0x00002b2f7ceff037 in PyImport_ExecCodeModuleEx (name=0x2b2f88809818 <PVAR_TIMER_idle+16> "", co=0x2b2f89076000, pathname=0x2b2f88809808 <PVAR_TIMER_idle> "") at Python/import.c:731
#36 0x00002b2f7cf01fd2 in load_source_module (name=0x2b2f88809818 <PVAR_TIMER_idle+16> "", pathname=0x2b2f89076000 "\001", fp=0x2b2f88809808 <PVAR_TIMER_idle>) at Python/import.c:1121
#37 0x00002b2f7cf03d8c in load_module (name=0x2b2f88809818 <PVAR_TIMER_idle+16> "", fp=0x2b2f89076000, pathname=0x2b2f88809808 <PVAR_TIMER_idle> "", type=-2005061824, loader=0x0) at Python/import.c:1928
---Type <return> to continue, or q <return> to quit---
#38 0x00002b2f7cf03c56 in load_package (name=<optimized out>, pathname=<optimized out>) at Python/import.c:1188
#39 load_module (name=0x2b2f88809818 <PVAR_TIMER_idle+16> "", fp=0x2b2f89076000, pathname=0x2b2f88809808 <PVAR_TIMER_idle> "", type=-2005061824, loader=0x0) at Python/import.c:1942
#40 0x00002b2f7cf0066f in import_submodule (mod=0x2b2f88809818 <PVAR_TIMER_idle+16>, subname=0x2b2f89076000 "\001", fullname=0x2b2f88809808 <PVAR_TIMER_idle> "") at Python/import.c:2725
#41 0x00002b2f7cf02e92 in load_next (mod=<optimized out>, altmod=0x4f, p_name=<optimized out>, buf=<optimized out>, p_buflen=<optimized out>) at Python/import.c:2539
#42 import_module_level (name=0x2b2f88809818 <PVAR_TIMER_idle+16> "", globals=0x2b2f89076000, locals=0x2b2f88809808 <PVAR_TIMER_idle>, fromlist=0x2b2f887d2f40 <MPID_nem_mem_region>, level=0)
    at Python/import.c:2247
#43 0x00002b2f7cf057b0 in PyImport_ImportModuleLevel (name=0x2b2f88809818 <PVAR_TIMER_idle+16> "", globals=0x2b2f89076000, locals=0x2b2f88809808 <PVAR_TIMER_idle>,
    fromlist=0x2b2f887d2f40 <MPID_nem_mem_region>, level=0) at Python/import.c:2312
#44 0x00002b2f7cecae7e in builtin___import__ (self=0x2b2f88809818 <PVAR_TIMER_idle+16>, args=0x2b2f89076000, kwds=0x2b2f88809808 <PVAR_TIMER_idle>) at Python/bltinmodule.c:49
#45 0x00002b2f7cdee38a in PyObject_Call (func=0x2b2f88809818 <PVAR_TIMER_idle+16>, arg=0x2b2f89076000, kw=0x2b2f88809808 <PVAR_TIMER_idle>) at Objects/abstract.c:2547
#46 0x00002b2f7ced8d3f in PyEval_CallObjectWithKeywords (func=0x2b2f88809818 <PVAR_TIMER_idle+16>, arg=0x2b2f89076000, kw=0x2b2f88809808 <PVAR_TIMER_idle>) at Python/ceval.c:4226
#47 0x00002b2f7ced4a5d in PyEval_EvalFrameEx (f=0x2b2f88809818 <PVAR_TIMER_idle+16>, throwflag=-1996005376) at Python/ceval.c:2628
#48 0x00002b2f7ced97ed in PyEval_EvalCodeEx (co=0x2b2f88809818 <PVAR_TIMER_idle+16>, globals=0x2b2f89076000, locals=0x2b2f88809808 <PVAR_TIMER_idle>, args=0x2b2f887d2f40 <MPID_nem_mem_region>,
    argcount=0, kws=0x4f, kwcount=0, defs=0x0, defcount=0, closure=0x0) at Python/ceval.c:3589
#49 0x00002b2f7ced92b4 in PyEval_EvalCode (co=0x2b2f88809818 <PVAR_TIMER_idle+16>, globals=0x2b2f89076000, locals=0x2b2f88809808 <PVAR_TIMER_idle>) at Python/ceval.c:669
#50 0x00002b2f7cf10c13 in run_mod (mod=<optimized out>, filename=<optimized out>, globals=<optimized out>, locals=<optimized out>, flags=<optimized out>, arena=<optimized out>) at Python/pythonrun.c:1376
#51 PyRun_StringFlags (str=0x2b2f88809818 <PVAR_TIMER_idle+16> "", start=-1996005376, globals=0x2b2f88809808 <PVAR_TIMER_idle>, locals=0x2b2f887d2f40 <MPID_nem_mem_region>, flags=0x0)
    at Python/pythonrun.c:1339
#52 0x00002b2f7cf10c99 in PyRun_SimpleStringFlags (command=0x2b2f88809818 <PVAR_TIMER_idle+16> "", flags=0x2b2f89076000) at Python/pythonrun.c:974
#53 0x00002b2f7cf29ce0 in Py_Main (argc=-2004838376, argv=0x2b2f89076000) at Modules/main.c:589
#54 0x00002b2f7dfc1c05 in __libc_start_main () from /lib64/libc.so.6
#55 0x00000000004012e9 in _start ()
```

**Comment by leonfoks** [comment]
@aragilar Sorry for the delay, I dropped the ball on this one...

I ran 
```
gdb --ex r --args python -c 'import h5py'
```

and did not get too much.

```
GNU gdb (GDB) Red Hat Enterprise Linux (7.2-83.el6)
Copyright (C) 2010 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>...
Reading symbols from /lustre/projects/software/anaconda/miniconda3/bin/python...done.
Starting program: /lustre/projects/software/anaconda/miniconda3/bin/python -c import\ h5py
[Thread debugging using libthread_db enabled]
--------------------------------------------------------------------------
It looks like orte_init failed for some reason; your parallel process is
likely to abort.  There are many reasons that a parallel process can
fail during orte_init; some of which are due to configuration or
environment problems.  This failure appears to be an internal failure;
here's some additional information (which may only be relevant to an
Open MPI developer):

  PMI2_Job_GetId failed failed
  --> Returned value (null) (14) instead of ORTE_SUCCESS
--------------------------------------------------------------------------
--------------------------------------------------------------------------
It looks like orte_init failed for some reason; your parallel process is
likely to abort.  There are many reasons that a parallel process can
fail during orte_init; some of which are due to configuration or
environment problems.  This failure appears to be an internal failure;
here's some additional information (which may only be relevant to an
Open MPI developer):

  orte_ess_init failed
  --> Returned value (null) (14) instead of ORTE_SUCCESS
--------------------------------------------------------------------------
--------------------------------------------------------------------------
It looks like MPI_INIT failed for some reason; your parallel process is
likely to abort.  There are many reasons that a parallel process can
fail during MPI_INIT; some of which are due to configuration or environment
problems.  This failure appears to be an internal failure; here's some
additional information (which may only be relevant to an Open MPI
developer):

  ompi_mpi_init: ompi_rte_init failed
  --> Returned "(null)" (14) instead of "Success" (0)
--------------------------------------------------------------------------
*** An error occurred in MPI_Init_thread
*** on a NULL communicator
*** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
***    and potentially your MPI job)
[igskahcmgslih04.cr.usgs.gov:98912] Local abort before MPI_INIT completed successfully; not able to aggregate error messages, and not able to guarantee that all other processes were killed!

Program exited with code 01.
Missing separate debuginfos, use: debuginfo-install glibc-2.12-1.209.el6_9.2.x86_64 libibumad-1.3.10.2.MLNX20150406.966500d-0.1.32101.x86_64 libibverbs-1.1.8mlnx1-OFED.3.2.0.1.2.32101.x86_64 libnl-1.1.4-2.el6.x86_64 librdmacm-1.0.21mlnx-OFED.3.1.1.5.5.32101.x86_64 opensm-libs-4.6.1.MLNX20160112.774e977-0.1.32101.x86_64 tcp_wrappers-libs-7.6-57.el6.x86_64
```

**Comment by dnck** [comment]
Hi, I'm just getting started with parallel computing with HDF5 and Python. I installed the Parallel version of HDF5 and I'm using openmpi as MPI. 

When I run 

```
import sys
import os
from mpi4py import MPI
import h5py
print h5py.version.info
```

I get the appropriate output, 

```

Summary of the h5py configuration
---------------------------------

h5py    2.7.1
HDF5    1.10.2
Python  2.7.12 |Anaconda custom (x86_64)| (default, Jul  2 2016, 17:43:17) 
[GCC 4.2.1 (Based on Apple Inc. build 5658) (LLVM build 2336.11.00)]
sys.platform    darwin
sys.maxsize     9223372036854775807
numpy   1.13.3
```

However, when I run, 


```
import sys
import os
import h5py
print h5py.version.info
```

I get the following error: 

```
[:59811] *** Process received signal ***
[:59811] Signal: Segmentation fault: 11 (11)
[:59811] Signal code: Address not mapped (1)
[:59811] Failing at address: 0x0
[:59811] [ 0] 0   libsystem_platform.dylib            0x00007fff63204f5a _sigtramp + 26
[:59811] [ 1] 0   ???                                 0x0000000000000000 0x0 + 0
[:59811] [ 2] 0   libopen-pal.20.dylib                0x00000001023d2a4f mca_base_framework_var_register + 79
[:59811] [ 3] 0   libopen-pal.20.dylib                0x00000001023d7828 mca_base_framework_register + 360
[:59811] [ 4] 0   libopen-pal.20.dylib                0x00000001023d78f4 mca_base_framework_open + 20
[:59811] [ 5] 0   libopen-pal.20.dylib                0x00000001023cf16a mca_base_framework_components_open + 218
[:59811] [ 6] 0   libmpi.20.dylib                     0x00000001021f8230 mca_pml_base_open + 304
[:59811] [ 7] 0   libopen-pal.20.dylib                0x00000001023d7989 mca_base_framework_open + 169
[:59811] [ 8] 0   libmpi.20.dylib                     0x000000010217d8b0 ompi_mpi_init + 1216
[:59811] [ 9] 0   libmpi.20.dylib                     0x0000000102199cd3 MPI_Init_thread + 99
[:59811] [10] 0   MPI.so                              0x00000001025a86e6 initMPI + 13478
[:59811] [11] 0   libpython2.7.dylib                  0x00000001000c2fda _PyImport_LoadDynamicModule + 170
[:59811] [12] 0   libpython2.7.dylib                  0x00000001000c1747 import_submodule + 327
[:59811] [13] 0   libpython2.7.dylib                  0x00000001000c194b load_next + 251
[:59811] [14] 0   libpython2.7.dylib                  0x00000001000c27df PyImport_ImportModuleLevel + 1231
[:59811] [15] 0   libpython2.7.dylib                  0x000000010009ce14 builtin___import__ + 132
[:59811] [16] 0   libpython2.7.dylib                  0x000000010000c761 PyObject_Call + 97
[:59811] [17] 0   libpython2.7.dylib                  0x000000010000cf42 call_function_tail + 98
[:59811] [18] 0   libpython2.7.dylib                  0x000000010000d3a4 PyObject_CallFunction + 212
[:59811] [19] 0   libpython2.7.dylib                  0x00000001000c2b71 PyImport_Import + 369
[:59811] [20] 0   _errors.so                          0x00000001006f4a94 __Pyx_ImportType + 52
[:59811] [21] 0   _errors.so                          0x00000001006f0f77 init_errors + 1479
[:59811] [22] 0   libpython2.7.dylib                  0x00000001000c2fda _PyImport_LoadDynamicModule + 170
[:59811] [23] 0   libpython2.7.dylib                  0x00000001000c1747 import_submodule + 327
[:59811] [24] 0   libpython2.7.dylib                  0x00000001000c1ae2 ensure_fromlist + 114
[:59811] [25] 0   libpython2.7.dylib                  0x00000001000c2934 PyImport_ImportModuleLevel + 1572
[:59811] [26] 0   libpython2.7.dylib                  0x000000010009ce14 builtin___import__ + 132
[:59811] [27] 0   libpython2.7.dylib                  0x000000010000c761 PyObject_Call + 97
[:59811] [28] 0   libpython2.7.dylib                  0x00000001000a1874 PyEval_CallObjectWithKeywords + 180
[:59811] [29] 0   libpython2.7.dylib                  0x00000001000a44ed PyEval_EvalFrameEx + 8317
[:59811] *** End of error message ***
Segmentation fault: 11
```

I think this is related to the already mentioned problem. Is there a solution I missed somehow?

**Comment by boegel** [comment]
I should add that using `export I_MPI_SLURM_EXT=0` before trying the `python -c "import h5py"` solved the problem for us, but this is of course specific to doing the `import` in an interactive SLURM session...

**Comment by timeu** [comment]
We are running into a similar issue. 
```
Summary of the h5py configuration
---------------------------------

h5py    2.8.0
HDF5    1.10.2
Python  3.6.6 (default, Jul 11 2019, 23:48:51) 
[GCC 7.3.0]
sys.platform    linux
sys.maxsize     9223372036854775807
numpy   1.15.0
```
When running `python -c "import h5py" in an interactive slurm session twice, the second call will cause following error:

```
[clip-c2-1:07270] PMIX ERROR: NOT-FOUND in file ptl_tcp.c at line 688
[clip-c2-1:07270] PMIX ERROR: NOT-FOUND in file ptl_usock.c at line 168
[clip-c2-1:07270] OPAL ERROR: Unreachable in file pmix2x_client.c at line 109
--------------------------------------------------------------------------
The application appears to have been direct launched using "srun",
but OMPI was not built with SLURM's PMI support and therefore cannot
execute. There are several options for building PMI support under
SLURM, depending upon the SLURM version you are using:

  version 16.05 or later: you can use SLURM's PMIx support. This
  requires that you configure and build SLURM --with-pmix.

  Versions earlier than 16.05: you must use either SLURM's PMI-1 or
  PMI-2 support. SLURM builds PMI-1 by default, or you can manually
  install PMI-2. You must then build Open MPI using --with-pmi pointing
  to the SLURM PMI library location.

Please configure as appropriate and try again.
--------------------------------------------------------------------------
*** An error occurred in MPI_Init_thread
*** on a NULL communicator
*** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
***    and potentially your MPI job)
[clip-c2-1:07270] Local abort before MPI_INIT completed completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!
```
This does not happen in a normal login shell (without srun)

**Comment by boegel** [comment]
I spent some time trying to fix this issue this afternoon, see #1544, but I'm a bit lost currently...

If anyone has some insight that would be helpful, I would appreciate it.

**Comment by takluyver** [comment]
I'm hoping that #1552 might have fixed this - does anyone want to try building h5py from master?

**Comment by boegel** [comment]
@takluyver I tested the fix in #1552 backported to h5py 2.10.0, and it works like a charm there.

**Comment by takluyver** [comment]
Right, thanks. I'm never quite sure with issues like that whether people have seem the same issue, or different ones with similar symptoms. But let's be optimistic and close this one - we can always reopen it or open a new issue if necessary.

**Comment by vasole** [comment]
Any idea about when 3.0 will be out? This fix is particularly critical.

**Comment by takluyver** [comment]
Not imminently, I'm afraid. There are still some fairly significant changes to land (e.g. #1338, #1526), and unfortunately most of the maintainers don't currently have a lot of time to spend on h5py. I'm doing some smaller changes by giving people a chance to object before I merge my own PRs, but I can't really do that for more in-depth changes.

@tacaswell has suggested that we do a bugfix release sooner. But I'm not keen on this idea - the release process itself is still fairly time consuming, and on top of that we'd have to decide which of the >100 merged PRs should be backported, and lots of people will want their particular fix to be included. So I'm worried that it would suck up whatever developer time is available and leave us in a swamp of competing priorities.

I would like to have smaller, more frequent releases of h5py in the future so we don't get into this situation again. But we're already well past the threshold for a small release, so first I think whatever time we can find has to be focused on merging changes and getting 3.0 out.

**Comment by Flamefire** [comment]
Some users on our cluster recently ran into this too and this consumed quite a lot of time to pinpoint. It's been a while since this fix landed and almost a year since the last release. Still it doesn't look like 3.0 is around the corner. May I recommend a strategy used by e.g. PyTorch:

- Open a branch going from the last release (2.10.0)
- Optional: Give a fixed (small) timeframe for a 2.10.1 bugfix release (e.g. 2 weeks or 1 month)
- Open an issue with announcement and for users to comment with PR numbers which are targeted on that branch, so let them do the backports
- Allow only bugfixes which are small and important enough such as this and already merged to master
- After the time expired do the release with everything merged so far.

This sounds like a good approach as it doesn't suck up much of maintainers time as all changes have already landed (in master) and it will fix pressing issues for users that care enough about that issue and this in a timely manner without requiring users to use the new major release.

**Comment by takluyver** [comment]
We're planning to spend some time in the next couple of weekends trying to tie up loose ends and get to 3.0.

I'm still not keen on bugfix releases for exactly the reasons I gave above: we barely have maintainer time to do *anything*, and the moment you say you're doing a bugfix release, you have to start making lots of difficult judgement calls about what should go into it. E.g. should support for HDF5 1.12 be backported? A lot of people will want it, but it's a fairly significant change which could conceivably cause problems.

Making bugfix releases also makes it easy to delay feature releases. Frankly, I'm worried that if we give in and start doing 2.10.x bugfixes, it will be another year before we get 3.0 out. There are real improvements that I want to get into a release soon.

PyTorch is in a very different situation, because there's a $100 billion company supporting its development and maintenance. There's a whole lot of things you can do when there are people consistently working on something, rather than trying to fit it in around their other responsibilities.

**Comment by Flamefire** [comment]
That's great to hear!

I'm aware of the time pressure and if 3.0 is close then it might indeed not be worth to do a bugfix release. Also yes PyTorch is different in terms of manpower, I just wanted to explain a potential process which I thought could be doable without spending much time. And about

> E.g. should support for HDF5 1.12 be backported? A lot of people will want it, but it's a fairly significant change which could conceivably cause problems.

That's a clear **no**  according to the outline policy of *bugfixes* only which are small and high reward. Nothing which can potentially break things or is complicated. The idea is to look at a PR and within seconds be able to merge or reject.

My point being: The current h5py is unusable in some situations. Namely (this bug) in an interactive slurm session:

- `srun --pty --nodes=1 --ntasks=1 bash` -> Get shell on 1 node
- `python -c `import h5py` --> hangs forever

This transitively affects other packages, e.g. TensorFlow.

That being said: Luckily we have means of applying the backported patch (via EasyBuild) so our users can use h5py once we reinstalled the module with the patch applied. But other users might still be blocked by this.

So thanks for your time answering this. I hope I made my motivation clear and as mentioned we have a way to workaround the issue. So it's up to you on how to proceed and I hope this information helps in any way on that decision.
